go/v1beta1/storage/memstore.go

```diff
@@ -0,0 +1,374 @@
+	for _, o := range occs {
+		clonedOccs = append(clonedOccs, proto.Clone(o).(*gpb.Occurrence))
+	}
+	occs = clonedOccs
+
+	errs := []error{}
+	created := []*gpb.Occurrence{}
+	for _, o := range occs {
+		occ, err := m.CreateOccurrence(ctx, pID, uID, o)
+		if err != nil {
+			// Occurrence already exists, skipping.
+			continue
+		} else {
+			created = append(created, occ)
+		}
+	}
+
+	return created, errs
+}
+
+// UpdateOccurrence updates the specified occurrence in memstore.
+func (m *memStore) UpdateOccurrence(ctx context.Context, pID, oID string, o *gpb.Occurrence, mask *fieldmaskpb.FieldMask) (*gpb.Occurrence, error) {
+	o = proto.Clone(o).(*gpb.Occurrence)
+
+	oName := name.FormatOccurrence(pID, oID)
+	m.Lock()
+	defer m.Unlock()
+	if _, ok := m.occurrencesByName[oName]; !ok {
+		return nil, errors.Newf(codes.NotFound, "Occurrence with name %q does not Exist", oName)
+	}
+
+	// TODO(#312): implement the update operation
+	o.UpdateTime = ptypes.TimestampNow()
+	m.occurrencesByName[oName] = o
+	return o, nil
+}
+
+// DeleteOccurrence deletes the specified occurrence in memstore.
+func (m *memStore) DeleteOccurrence(ctx context.Context, pID, oID string) error {
+	oName := name.FormatOccurrence(pID, oID)
+	m.Lock()
+	defer m.Unlock()
+	if _, ok := m.occurrencesByName[oName]; !ok {
+		return errors.Newf(codes.NotFound, "Occurrence with name %q does not Exist", oName)
+	}
+	delete(m.occurrencesByName, oName)
+	return nil
+}
+
+// GetNote gets the spcified note from memstore.
+func (m *memStore) GetNote(ctx context.Context, pID, nID string) (*gpb.Note, error) {
+	nName := name.FormatNote(pID, nID)
+	m.RLock()
+	defer m.RUnlock()
+	n, ok := m.notesByName[nName]
+	if !ok {
+		return nil, errors.Newf(codes.NotFound, "Note with name %q does not Exist", nName)
+	}
+	return n, nil
+}
+
+// ListNotes returns up to pageSize number of notes for the project pID beginning
+// at pageToken, or from start if pageToken is the empty string.
+func (m *memStore) ListNotes(ctx context.Context, pID, filter, pageToken string, pageSize int32) ([]*gpb.Note, string, error) {
+	ns := []*gpb.Note{}
+	m.RLock()
+	defer m.RUnlock()
+	for _, n := range m.notesByName {
+		if strings.HasPrefix(n.Name, fmt.Sprintf("projects/%v", pID)) {
+			ns = append(ns, n)
+		}
+	}
+	sort.Slice(ns, func(i, j int) bool {
+		return ns[i].Name < ns[j].Name
+	})
+	startPos := parsePageToken(pageToken, 0)
+	endPos := min(startPos+int(pageSize), len(ns))
+	return ns[startPos:endPos], nextPageToken(endPos, len(ns)), nil
+}
+
+// CreateNote creates the specified note in memstore.
+func (m *memStore) CreateNote(ctx context.Context, pID, nID string, uID string, n *gpb.Note) (*gpb.Note, error) {
+	n = proto.Clone(n).(*gpb.Note)
+	nName := name.FormatNote(pID, nID)
+
+	m.Lock()
+	defer m.Unlock()
+	if _, ok := m.notesByName[nName]; ok {
+		return nil, errors.Newf(codes.AlreadyExists, "Note with name %q already exists", n.Name)
+	}
+
+	n.Name = nName
+	n.CreateTime = ptypes.TimestampNow()
```

Comment:
I think update time is supposed to start at create time.

Reply:
Good point!